File: 2050-count-good-numbers/2050-count-good-numbers.cpp
class Solution {
public:
    
    const int M = 1e9+7;
    long long Expo(long long a, long long b){ //a^b
        if(b==0){
            return 1;
        }
        long long half = Expo(a,b/2);
        long long result = (half * half)%M;
        if(b%2==1){
            result = (result*a)%M;
        }
        return result;
    }

    long long countGoodNumbers(long long n) {
        long long even = Expo(5,(n+1)/2);
        long long odd = Expo(4,n/2);

        return (even * odd)%M;
    }
};

File: 2050-count-good-numbers/2050-count-good-numbers.java
class Solution {
    private static final long MOD = 1000000007;

    public int countGoodNumbers(long n) {
        long evens = (n + 1) / 2;
        long odds = n / 2;

        long part1 = modPow(5, evens, MOD); 
        long part2 = modPow(4, odds, MOD); 

        return (int)((part1 * part2) % MOD);
    }

    long modPow(long base, long exp, long mod) {
        long result = 1;
        while (exp > 0) {
            if (exp % 2 == 1) { // if exp is odd
                result = (result * base) % mod;
            }
            base = (base * base) % mod; 
            exp = exp / 2;              
        }
        return result;
    }

}
